### PR TITLE
fix(DDAPP-12052): add additional checks and time out timer in play path

### PR DESCRIPTION
**Pull request by johanvanderboom-tomtom**
# Description

Ensure the callback is called (when present) for all paths in play() and add a timer as a fail safe to ensure completion will be done if completion callback is not called so that the higher layer can do a proper clean up in all cases. (See the [PR](https://github.com/tomtom-internal/amigo-mobile/pull/4848) for the changes required on the AmiGO side)

## Context

|  |  |
| ----------------- | -------- |
| **Type**          | Bugfix |
| **Ticket**        | DDAPP-12052 |

## Implementation description

For all paths in play() that result in not playing the audio the callback is called with a failure result so that the Audio Source Manager can ensure cleanup.

As an additional fail safe a timer is started when playback is started which when fired (audio completion callback is not called) will call the play() callback so that the Audio Source Manager can clean up. This is added as even the paths where the callback was not called, they would have resulted in a error thrown and a crash which is not seen in BugSnag. This timer is added to see if it actually occurs that the audio completion callback is not called (the Audio Source Manager will notify BugSnag when this hapens)

## Test Plan

Test together with the changes for this in AmiGo

## Analytics

n/a

# Checklist (reminder)

- [x] The PR has a clear, descriptive title
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Unit tests are added to test expected and unexpected behaviour
- [ ] If there are strings that need translating, I have triggered the [Localization Process](https://confluence.tomtomgroup.com/display/DDAPP/Localization+Process)
- [x] Labels for your team are added to the PR

For PRs that change the UI:

No UI changes

# For the reviewers

Copy-paste this in a comment below, post it, and start checking things.

<details>
<summary>Reviewer Comment</summary>

```
### Dependencies

- [ ] I can verify the new dependency added in the `package.json` does not seem malicious.

### Review

- [ ] I understand the purpose of this PR
- [ ] I understand the code of this PR
- [ ] Works correctly on Android
- [ ] Works correctly on iOS
- [ ] The code quality is high (proper location, proper structure, clear names...)
- [ ] There's enough documentation to explain what is going on.

#### Just before merging

- [ ] All my requested changes have been resolved
- [ ] The branch is in sync with master
```
</details>

### CODE REVIEW

File: android/src/main/java/com/zmxv/RNSound/RNSoundModule.java [hunk 1]
```diff
@@ -349,11 +370,14 @@ public synchronized void onCompletion(MediaPlayer mp) {
...
-        if (callbackWasCalled) return true;
+        if (callbackWasCalled) {
+          return true;
+        }
+        cancelTimer(key);
         callbackWasCalled = true;
         try {
-          callback.invoke(true);
-        } catch (Exception e) {
+          callback.invoke(PLAY_RESULT_SUCCESS);
```
Comment (casparvanderheiden-tomtom) [inline android/src/main/java/com/zmxv/RNSound/RNSoundModule.java]: Should this be `PLAY_RESULT_SUCCESS` in onError?

**Review by casparvanderheiden-tomtom — COMMENTED** [review]
One question about `RnSoundModule.onError` (Android) It seems strange to call the callback with `PLAY_RESULT_SUCCESS` there. I saw that in the original code they called the callback with `true` (Which, if I understand correctly, is the value for success) but that might be an error on their part?

File: android/src/main/java/com/zmxv/RNSound/RNSoundModule.java (hunk 1 above)
Comment (johanvanderboom-tomtom) [reply android/src/main/java/com/zmxv/RNSound/RNSoundModule.java]: changed to `PLAY_RESULT_FAILURE` in commit [#0423746](https://github.com/tomtom-international/react-native-sound/pull/5/commits/04237469ad387c6797d4f7cb7eff742a99fb60ef)

**Comment by johanvanderboom-tomtom** [comment]
> One question about `RnSoundModule.onError` (Android) It seems strange to call the callback with `PLAY_RESULT_SUCCESS` there. I saw that in the original code they called the callback with `true` (Which, if I understand correctly, is the value for success) but that might be an error on their part?

 I agree that the onError should return a fail in the callback as it is supposed to be an error in playback

**Review by casparvanderheiden-tomtom — APPROVED** [review]
### Dependencies

- [ ] I can verify the new dependency added in the `package.json` does not seem malicious.

### Review

- [x] I understand the purpose of this PR
- [x] I understand the code of this PR
- [ ] Works correctly on Android (not tested)
- [ ] Works correctly on iOS (not tested)
- [x] The code quality is high (proper location, proper structure, clear names...)
- [x] There's enough documentation to explain what is going on.

#### Just before merging

- [x] All my requested changes have been resolved
- [x] The branch is in sync with master

**Pull request merged**